lms/services/canvas_api/client.py

```diff
@@ -250,22 +260,90 @@ def list_files(self, course_id, sort="position"):
                     "lms_id": file["id"],
                     "name": file["display_name"],
                     "size": file["size"],
+                    "parent_lms_id": file["folder_id"],
                 }
                 for file in files
             ]
         )
 
+        if self._with_folders:
+            folders = self._list_folders(course_id)
```

Comment:
We can get all the folders in one API call.

Reply:
The documentation at https://canvas.instructure.com/doc/api/files.html#method.folders.list_all_folders says that the API call is paginated, but it looks like the low-level client here handles following pagination links automatically?

Do we have any idea of the largest number of files / folders that we've seen so far is?

Reply:
This is probably worth looking into, I think some courses do have very large numbers of files

Reply:
There are few courses with over a thousand files. We can't measure the number of folders but we can guess that the worst case scenario would be around the same order of magnitude or probably less than the files ones.

The average number of files is 45 per course.


```
select application_instance_id, course_id, count(*) as per_course from file where type = 'canvas_file'
group by application_instance_id, course_id
order by per_course desc
``` 


Is this going to take some of the request over the timeout limit? Hard to tell. Maybe.

We could:

- Redo files support to work like blackboard's, query one level at time. This will probably be better in the general case but there might still be edge cases with loads of files at the root level or simlar.

We'd need new scopes for this so we'll need to support two different files implementations and one for folders.

- Sprinkle some async. Get folders and files simultaneously? We might get some performance benefits but our use of async feels a bit like a hack.


- Re-architect some of the app to be able to do HTTP requests in the background and communicate the results to the backend.

The LMS app is becoming (always has been?) glue around third party API, I reckon this is the way to go in the long term.

My preference here would be to deploy this as it is (is behind a feature flag anyway) and do a serious commit to implement the last option to be able do HTTP requests in the background.

We talked in the past about how that option will look:

- The backend will queue celery task for different features. This is pretty straightforward.

- The frontend will need to poll/ use websockets to get the results.

- BE/FE will need to agree on an API to serialize not only the results but errors, access token requests and refresh. We could replicate the same API we have now for direct HTTP requets in the results API.

Reply:
> My preference here would be to deploy this as it is (is behind a feature flag anyway)

Do we have any limits in place to bound the amount of time we spend fetching files/folders? I think it probably does make sense to get this deployed. If there is a way we can easily impose a sane upper limit on the size of the file tree that might be wise to do.